torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -729,6 +755,104 @@ void Reducer::finalize_backward() {
   local_used_maps_reduced_ = false;
 }
 
+void Reducer::sync_bucket_indices(
+    std::vector<std::vector<size_t>>& bucket_indices) {
+  int64_t broadcast_bucket_size = DEFAULT_BROADCAST_BUCKET_BYTES;
+
+  auto num_buckets = bucket_indices.size();
+  std::vector<size_t> bucket_sizes;
+  int64_t total_size = 0;
+  for (size_t i = 0; i < num_buckets; i++) {
+    auto bucket_size = bucket_indices.at(i).size();
+    bucket_sizes.push_back(bucket_size);
+    total_size += bucket_size;
+  }
+
+  at::TensorOptions options;
+  options = options.dtype(at::kInt);
+  options = options.device(replicas_[0][0].device());
+
+  // Group indices and num_bucket together into indices_tensor
+  // Broadcast this tensor first, as its size is equal among all processes
+  auto indices_tensor = at::empty({total_size + 1}, at::kInt);
+  auto indices_accessor = indices_tensor.accessor<int, 1>();
+  auto indices_accessor_Index = 0;
+  for (size_t i = 0; i < num_buckets; i++) {
+    const auto& bucket_size = bucket_indices.at(i).size();
+    for (size_t j = 0; j < bucket_size; j++) {
+      indices_accessor[indices_accessor_Index++] = bucket_indices[i][j];
+    }
+  }
+  indices_accessor[indices_accessor_Index] = num_buckets;
+
+  // Copy CPU tensor to device tensor, as the process_group_ could be NCCL and
+  // it can only broadcast device tensors.
+  auto indices_tensor_device = at::empty({total_size + 1}, options);
+  indices_tensor_device.copy_(indices_tensor, true);
+  broadcast_coalesced(
```

Comment:
Since `indices_tensor_device` is a single tensor, we don't need "coalesced" here? Looking at its implementation in `comm.cpp`, it is calling `compute_bucket_assignment_by_size` again and have additional unnecessary for/while loops. Can we directly call broadcast and wait on the result here?

Reply:
sounds good! let me try that